File: Suraj/01 Sorting/Quick_sort.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int partitionIndex(vector<int> &nums,int low ,int high){
        int i = low;
        int j = high;
        int pivot = nums[low];
        while(i<j){
            while(nums[i] <= pivot && i<high) i++;
            while(nums[j]>pivot && j>low) j--;
            if(i<j)swap(nums[i],nums[j]);

        }
        swap(nums[low],nums[j]);
        return j;
    }
    void qs(vector<int> &nums,int low ,int high){
        if(low<high){
            int pIndex = partitionIndex(nums,low,high);
            qs(nums,low,pIndex-1);
            qs(nums,pIndex+1,high);
        }
    }
    vector<int> quickSort(vector<int>& nums) {
        int n = nums.size();
        qs(nums,0,n-1);
        return nums;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {3, 1, 4, 1, 5, 9, 2, 6, 5};
    vector<int> res = sol.quickSort(nums);
    for (int i = 0; i < res.size(); i++) {
        cout << res[i] << " ";
    }
    return 0;
}

File: Ansh/sorting/quickSort.py
class Solution:
    def quickSort(self, nums):
        self.qs(nums,0,len(nums)-1)
        return nums
    
    def qs(self,nums,low,high):
        if(low<high):
            pIndex=self.partition(nums,low,high)
            self.qs(nums,low,pIndex-1)
            self.qs(nums,pIndex+1,high)
    
    def partition(self,nums,low,high):
        i=low
        j=high
        pivot=nums[low]
        while(i<j):
            while(nums[i]<=pivot and i<=high-1):
                i+=1
            
            while(nums[j]>pivot and j>=low+1):
                j-=1
            
            if(i<j):
                nums[i],nums[j]=nums[j],nums[i]
        nums[low],nums[j]=nums[j],nums[low]
        return j
    
if __name__=="__main__":
    nums=[64, 25, 12, 22, 11]
    s=Solution()
    print(s.quickSort(nums))
    # by ANSH PATERIYA